### PR TITLE
[Serializer] Remove some type hints for API Platform compatibility

### LINKED ISSUE #43021: [META] Symfony 6 return type declarations
Symfony 6 comes (almost) fully typed. If you're an open source package maintainer, adding a return type might cause BC breaks (as it forces your users to add the type as well if they implement or override the method).
We might consider, _on a case by case basis_, whether it's worth to postpone adding a type declaration for a specific method to Symfony 7. This might help you, as a maintainer, to release a new minor instead of a new major version to add support for Symfony 6.

1. In order to properly judge this, we ask you to **install 5.4-dev in your package and run the [type patch utility](https://symfony.com/doc/5.4/setup/upgrade_major.html#upgrading-to-symfony-6-add-native-return-types)** twice (also temporarily install the `symfony/error-handler` component if it's not available already):

   ```
   $ SYMFONY_PATCH_TYPE_DECLARATIONS=force=1 ./vendor/bin/patch-type-declarations
   $ SYMFONY_PATCH_TYPE_DECLARATIONS=force=1 ./vendor/bin/patch-type-declarations
   ```

2. This script patches any methods that are private, final, `@internal` or tests. **Please double check if the unpatched methods should not be final or internal** (i.e. should a user really extend or use this class?).

3. If the _second_ run **produces any deprecations in the output for methods that must be considered public/protected, please copy paste them as a comment to this issue** (including a link to the package). E.g:

   > https://github.com/wouterj/WouterJEloquentBundle
   > ```
   > WouterJ\EloquentBundle\Command\SeedCommand
   > Method "Symfony\Component\Console\Command\Command::execute()" might add "int" as a native return type declaration in the future. Do the same in child class "WouterJ\EloquentBundle\Command\SeedCommand" now to avoid errors or add an explicit @return annotation to suppress this message.
   > ```

   Note that the current 6.0 branch reflects which types are not going to be included in that release. These still produce deprecations in 5.4+, which can be resolved by adding the `@return` PHPdoc.

   (please only post deprecations mentioning classes from Symfony, other deprecations should be reported in the relevant package issue tracker)

We don't promise anything, but we might postpone some types based on the feedback in *this* issue. For instance, we already did for the `Command::execute(): int` method shown in the example.

### PR DESCRIPTION
**Pull request by PierreRebeilleau**
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #43021, https://github.com/api-platform/core/issues/4495
| License       | MIT

Since Symfony 6.0 and the addition of type hinting, API Platform's tests fails. After a talk with @dunglas , we decide to remove some type hinting in the serializer component in order for api-platform to work correctly.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Encoder/DecoderInterface.php
Comment (stof) [inline src/Symfony/Component/Serializer/Encoder/DecoderInterface.php]: Please keep the empty phpdoc lines separating blocks.

File: src/Symfony/Component/Serializer/Encoder/ChainDecoder.php
Comment (stof) [inline src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: I would keep the return type on that final method, making it ready for when the interface actually adds the type.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php
Comment (stof) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Do we actually need to remove the argument type here ? Child class are allowed to have a wider parameter type (and so are allowed to remove the type)

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (stof) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: is the `string` type really an issue here ?

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (stof) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: same here

**Comment by wouterj** [comment]
Can you please provide a bit more context? This PR removes a lot of typehints: are they all needed to remove? And why are they required to be removed?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (stof) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: any reason to remove the return type on those implementation ?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: why reverting argument's type?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: how related is this?

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: should be reverted

**Review by stof — COMMENTED** [review]
Note that the approach about adding return type in subclasses to be compatible with Symfony 5.4 and 6.0 at the same time relies on the partial covariance support of PHP 7.2 (partial because it works only for parent methods without an explicit type).
So ApiPlatform will need to require PHP 7.2+ to be able to write code supporting Symfony 6

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php
Comment (stof) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: this should indeed all be reverted. Child classes are allowed to widen the argument type, so this should not be an issue for ApiPlatform.

**Comment by OskarStark** [comment]
@stof PHP 6 -> Sf6 ?

**Comment by stof** [comment]
@OskarStark indeed. Edited to fix it.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php
Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: API Platform supports PHP 7.1, so we cannot use union types. Or we'll have to bump to 7.2 as already mentioned, which is something we can do.

**Comment by aurimasniekis** [comment]
I am very curious what kind of issues these added types cause because maybe they can involve other projects too? Because from the looks I only can suspect the issue with `mixed` type but thinks like returns bool doesn't really make sense to cause issues.

**Comment by dunglas** [comment]
API Platform v2 needs this to be able to support both Symfony 4.4 and Symfony 6.
API Platform v2 supports PHP 7.1, this means no (partial) covariance/contravariance, no union types, and no `mixed`.

In API Platform 3 (to be released soon), we'll support only PHP 8.1 and Symfony 6, but having support for Symfony 6 in API Platform 2 will ease the upgrade process.

**Comment by aurimasniekis** [comment]
Wouldn't it make sense to limit API Platform v2 to Symfony 5.4 only? Because I feel like this is stopping Symfony6 from going full PHP8?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: should be reverted

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: mixed should be added back

**Comment by stof** [comment]
@dunglas if you cannot bump to PHP 7.2+, the whole migration path prepared for Symfony 6 types won't work for you, due to not having the necessary prerequisite. I would suggest bumping to PHP 7.2+ as part of that effort. Not supporting partial covariance means that you will need us to revert all type additions done in Symfony for methods that you implement. That's too strong as a requirement IMO (especially as we are in progress of doing the stable release)
Side note: this need for partial covariance to add types is the main reason why Symfony 5 requires PHP 7.2+.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: mixed should stay

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: the union should stay

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: mixed

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: mixed

File: src/Symfony/Component/Serializer/Normalizer/ContextAwareDenormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ContextAwareDenormalizerInterface.php]: mixed

File: src/Symfony/Component/Serializer/Normalizer/ContextAwareNormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ContextAwareNormalizerInterface.php]: mixed

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: stays

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php]: mixed stays

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php]: mixed stays

**Review by nicolas-grekas — COMMENTED** [review]
please revert everything that is not related to return types

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (PierreRebeilleau) [reply src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php]: Thanks for the review, sorry for making you write endless "mixed" :pray:

**Comment by stof** [comment]
To me, all argument types should stay, and all return types supported on PHP 7.2 should stay as well.
Using PHP 7.2+ is a prerequisite of our approach about adding types (thanks to the partial variance rules supported in PHP 7.2) and those rules mean that argument types won't need to be added in the ApiPlatform child classes (so even union types are not an issue).

**Comment by dunglas** [comment]
@stof As shown by our test suite, these types are the only ones preventing supporting PHP 7.1.
We discussed this privately with @nicolas-grekas a while ago, and we agreed that having support for Symfony 6 in API Platform 2 would be nice. But we can also consider dropping support for PHP 7.1.

**Comment by stof** [comment]
@dunglas given that 6.0 is tagged, this is too late for this PR to make it in anyway (feedback on the type migration was asked in September, and this PR was opened while the release process for the stable release was already started)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: boo :)

**Comment by nicolas-grekas** [comment]
I'm personally still fine to consider this change for 6.0.1. It's not like we'll break a lot of code by merging this.

Drupal has been asking for relaxing the return type before on `NormalizerInterface`, so it's not like we don't have any open discussion on the topic.

And smoothing the migration for ApiPlatform looks worth it to me.

**Comment by dunglas** [comment]
@stof I understand that this PR comes (too) late, but most contributors are benevolent and we do what we can with our limited workforce. We started working on this a while ago, we already made the whole codebase compatible with Symfony 5.4 back in October and we discussed the approach thoroughly (mostly off-GitHub) with @chalasr, @soyuka, and @nicolas-grekas.

If we have no other option, we'll make API Platform 2 compatible only with Symfony 5.4, and API Platform only with Symfony 6, but I think that it will also have a negative impact on Symfony, as all projects using API Platform will not be able to upgrade to Symfony 6 until they upgrade to PHP 8.1 and API Platform 3 (not mentioning issues impacting Drupal).

File: src/Symfony/Component/Serializer/Serializer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: `final`, so not needed?

File: src/Symfony/Component/Serializer/Encoder/CsvEncoder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/CsvEncoder.php]: does apiplatform extend CsvEncoder?
if not, this change is not needed isn't it?
can you please consider the same question for all concrete classes?

Comment (dunglas) [reply src/Symfony/Component/Serializer/Encoder/CsvEncoder.php]: We don't. Let’s keep the types for concrete classes we don't extend.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (jderusse) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: This method is not extends in ApiPlatform why changing everything?

**Review by jderusse — COMMENTED** [review]
I'm sad about this step backward.

What's about fixing that in ApiPLatform by providing 2 implementations (like we do for ages, for instance in the PhpUnitBridge)?

```php
// JsonEncoder.php
final class JsonEncoder implements EncoderInterface, DecoderInterface
{
  use EncoderTrait;

  private function _decode($data, $format, array $context = [])
 {...}
}

// EncoderTrait.php
if (\PHP_VERSION_ID >= 70200) {
  trait EncoderTrait
  {
    use EncoderTrait72;
  }
} else {
  trait EncoderTrait
  {
    use EncoderTrait71;
  }
}

// EncoderTrait72.php
trait EncoderTrait72
{
  public function decode($data, $format, array $context = []): mixed
  {
    return $this->_decode($data, $format, $context);
  }
}

// EncoderTrait71.php
trait EncoderTrait71
{
  public function decode($data, $format, array $context = [])
  {
    return $this->_decode($data, $format, $context);
  }
}
```

**Review by derrabus — CHANGES_REQUESTED** [review]
We should keep the changes to a minimum here. I'm open to removing return types from interfaces. But this PR proposes to revert them from implementations and I'd really like to hear a very good reason for doing so. Also, it should not be necessary to remove any parameter types.

**Comment by PierreRebeilleau** [comment]
> I'm sad about this step backward.
> 
> What's about fixing that in ApiPLatform by providing 2 implementations (like we do for ages, for instance in the PhpUnitBridge)?
> 
> ```
> // JsonEncoder.php
> final class JsonEncoder implements EncoderInterface, DecoderInterface
> {
>   use EncoderTrait;
> 
>   private function _decode($data, $format, array $context = [])
>  {...}
> }
> 
> // EncoderTrait.php
> if (\PHP_VERSION_ID >= 70200) {
>   trait EncoderTrait
>   {
>     use EncoderTrait72;
>   }
> } else {
>   trait EncoderTrait
>   {
>     use EncoderTrait71;
>   }
> }
> 
> // EncoderTrait72.php
> trait EncoderTrait72
> {
>   public function decode($data, $format, array $context = []): mixed
>   {
>     return $this->_decode($data, $format, $context);
>   }
> }
> 
> // EncoderTrait71.php
> trait EncoderTrait71
> {
>   public function decode($data, $format, array $context = [])
>   {
>     return $this->_decode($data, $format, $context);
>   }
> }
> ```

I talked to kevin about it and the API Platform's core team does not agree with this solution, it would change too many things on the API Platform

**Comment by wouterj** [comment]
I personally very unsure about these changes. This is the most significant revert that is proposed for removing type hints.

We've been quite strict about which types to revert or not, even deciding against reverting the typehint of the `NormalizerInterface` when Drupal proposed this. Question to the core team: Are the arguments made back then no longer valid? Or what circumstances have changed to not allow it 2 months ago, but allow it now?

Then, and I admit that this is not technical but purely political, we've been especially early (3 - 4 months prior to the stable release) with announcing these changes and making as much noise as possible about it to gather feedback from open source communities relying on Symfony. I'm very unsure about accepting the most significant changes after the stable release.
That is, unless we accept the policy that removing return type declarations is considered a bugfix and we will continue to revert return types if they are reported in the next 2 years. I don't think this is a great policy, as this means applications can not rely on the return type declarations - removing all type safety that strict typing offers.

At last, if I read the discussion above correctly, there are changes that can fix this on APIP's side instead of on Symfony's side. The easiest solution of all removing support for PHP 7.1 in APIP 2. Another solution being using 2 implementations, which arguably is less nice. Given other communities didn't need these changes (or didn't test with Symfony 6 yet?), I feel like the changes are specific to APIP. Fixing this on APIP's side seems to be reasonable in that case.

---

*quick note: I'm very unsure, but I won't block this merge - which is why I left a plain PR comment. But I think it's good to not take this too lightly.*

**Comment by fabpot** [comment]
Looks like I’ll be the one blocking this merge for all the reasons mentioned in the comments. :-1:

**Review by nicolas-grekas — APPROVED** [review]
Now that the PR has been restricted to what is really required, it makes sense to me.

Removing return types from abstract classes and interfaces that are designed to be extended/implemented by userland/libs is exactly why we opened #43021, and why we already reverted some return types from 6.0.

*Not* having these return types creates zero issues for us and gives more time to the community to adopt those return types. Yes, it's less strict. But I don't think strictness is a good driver when there is a clear advantage to being pragmatic here.

:+1: on my side since the pros/cons balance is in favor of the pros to me.

**Review by dunglas — APPROVED** [review]

**Review by stof — APPROVED** [review]

**Comment by dunglas** [comment]
@jderusse we thought about that but as API Platform 2 will be EOL soon and that we already have to maintain 2 implementations in the 2.7 branch (the legacy one, and the one compatible with API Platform 3), this increases the maintenance burden. Also, we have a lot of classes implementing these interfaces.

**Review by chalasr — APPROVED** [review]
Although it's super late, this does not hurt. Let's help each other.

**Comment by wouterj** [comment]
I still fail to understand how we forced a lot of work on Drupal for the `NormalizerInterface` 2 months ago, only to revert the return type after a stable release. I feel very sorry for Drupal for all the time they had to spent to rewrite their serializer implementations.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (derrabus) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:138]: The return type has to be re-added to the PHPDoc block, here and for all other methods changed in this file.

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (derrabus) [inline src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php:39]: Please restore `@return`!

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (derrabus) [inline src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php:49]: Please restore `@return`!

**Comment by nicolas-grekas** [comment]
> I still fail to understand how we forced a lot of work on Drupal for the NormalizerInterface 2 months ago, only to revert the return type after a stable release. I feel very sorry for Drupal for all the time they had to spent to rewrite their serializer implementations.

On my side: because of the reason why. Reverting the return type for BC/FC purpose is legit. Reverting for allowing back a hack is certainly not. Aka in this PR, we know we'll be able to add the return type in sf7 because APIP will have released v3. On the Drupal case, we don't know that, quite the opposite.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:138]: we don't have this one on 5.4 so it's not needed (same namespace makes inheritdoc enough)

File: src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Normalizer/NormalizerInterface.php:49]: good catch with this one!

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php
Comment (derrabus) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:138]: All right.

**Review by derrabus — APPROVED** [review]

**Comment by stof** [comment]
Drupal was not asking to delay the addition of the native return type. It was asking to change the contract of the interface (because native return type are enforced and so they could not violate the contract anymore).

**Comment by wouterj** [comment]
:ok_hand:  alright, makes sense.

**Review by stof — APPROVED** [review]

**Review by fabpot — APPROVED** [review]
The last iteration makes more sense to me. Let's do it.

**Comment by nicolas-grekas** [comment]
Thank you @PierreRebeilleau.

**Pull request merged**
